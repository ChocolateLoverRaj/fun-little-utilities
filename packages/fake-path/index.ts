export const symCwd = Symbol('cwd')
export const symRoot = Symbol('root')

type StringKey<Object> = {
  [key in string & keyof Object]: Object[key]
}

export type Path = StringKey<FakePath>

const isEmpty = (path: string) => path === '.' || path === ''
const normalizeEmptyPath = (path: string) => isEmpty(path) ? '.' : path

export abstract class FakePath {
  public abstract readonly [symCwd]: string
  public abstract readonly [symRoot]: readonly string[]
  public abstract readonly sep: string

  public readonly isAbsolute = (path: string): boolean =>
    this[symRoot].some(root => path.startsWith(root))

  public readonly join = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.join(...rest)
    if (isEmpty(head)) return tail
    if (isEmpty(tail)) return head
    if (tail === '..') return this.dirname(head)
    if (tail.startsWith('..' + this.sep)) {
      return this.join(this.dirname(head), tail.slice(('..' + this.sep).length))
    }
    const left = head.endsWith(this.sep) ? head.slice(0, -1) : head
    const right = tail.startsWith(this.sep) ? tail.slice(1) : tail
    return this.normalize(left + this.sep + right)
  }

  public readonly resolve = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.resolve(...rest)
    if (this.isAbsolute(tail)) return tail
    return this.join(this[symCwd], head, tail)
  }

  public readonly dirname = (path: string): string => {
    if (path === '' || path === '.' || path === this.sep) return path
    const segments = path.split(this.sep).slice(0, -1)
    if (!segments.length) return '.'
    return this.normalize(this.join(...segments))
  }

  public readonly basename = (path: string): string => {
    const segments = path.split(this.sep)
    return this.normalize(segments[segments.length - 1])
  }

  public readonly normalize = (path: string): string => {
    if (isEmpty(path)) return '.'
    if (path.startsWith(this.sep)) {
      const tail = this.normalize(path.slice(this.sep.length))
      return this.sep + tail
    }
    const segments = path
      .split(this.sep)
      .filter(x => !isEmpty(x))
    // tslint:disable-next-line:one-variable-per-declaration
    for (let i = 0, j = 1; j < segments.length; ++i, ++j) {
      if (segments[j] === '..') {
        segments[i] = segments[j] = ''
      }
    }
    const result = segments.filter(Boolean).join(this.sep)
    return normalizeEmptyPath(result)
  }
}

export default FakePath
