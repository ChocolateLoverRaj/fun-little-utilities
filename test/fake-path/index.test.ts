import { partition } from '@tsfun/array'
import { FakePath, symCwd, symRoot } from 'fake-path'

class Path extends FakePath {
  public readonly [symCwd] = '/working/directory'
  public readonly [symRoot] = ['/', '\\', 'C:\\', 'D:\\']
  public readonly sep = '/'
}

it('isAbsolute', () => {
  const { isAbsolute } = new Path()
  const [absolute, relative] = partition([
    '/',
    'C:\\',
    '/a/b/c',
    'C:\\a\\b',
    '.',
    '..',
    'a/b/c',
    './a/b',
    '../a/b'
  ], isAbsolute)
  expect({ absolute, relative }).toMatchSnapshot()
})

describe('join', () => {
  it('relative only', () => {
    const { join } = new Path()
    expect(join('a/b', 'c/d/e', 'f', '.', 'g', '', 'h')).toBe('a/b/c/d/e/f/g/h')
  })

  it('with ".." somewhere', () => {
    const { join } = new Path()
    expect(join('a', 'b/../c', 'd/./e/f', '../g', 'h/i/..', 'j', '../.', './k', './..'))
      .toBe('a/c/d/e/g/h')
  })

  it('first argument is ""', () => {
    const { join } = new Path()
    expect(join('', 'a', 'b', 'c')).toBe('a/b/c')
  })

  it('first argument is "."', () => {
    const { join } = new Path()
    expect(join('.', 'a', 'b', 'c')).toBe('a/b/c')
  })

  it('first argument is ".."', () => {
    const { join } = new Path()
    expect(join('..', 'a', 'b', 'c')).toBe('../a/b/c')
  })

  it('first argument starts with "./"', () => {
    const { join } = new Path()
    expect(join('./a/b', 'c')).toBe('a/b/c')
  })

  it('first argument starts with "../"', () => {
    const { join } = new Path()
    expect(join('../a/b', 'c')).toBe('../a/b/c')
  })

  it('second argument is ""', () => {
    const { join } = new Path()
    expect(join('a/b/c', '')).toBe('a/b/c')
  })

  it('second argument is "."', () => {
    const { join } = new Path()
    expect(join('a/b/c', '.')).toBe('a/b/c')
  })

  it('second argument is ".."', () => {
    const { join } = new Path()
    expect(join('a/b/c', '..')).toBe('a/b')
  })

  it('second argument ends with "/"', () => {
    const { join } = new Path()
    expect(join('a/b', 'c/')).toBe('a/b/c')
  })

  it('second argument ends with "/."', () => {
    const { join } = new Path()
    expect(join('a/b', 'c/.')).toBe('a/b/c')
  })

  it('second argument starts with "../"', () => {
    const { join } = new Path()
    expect(join('a/b', '../c')).toBe('a/c')
  })

  it('second argument ends with "/.."', () => {
    const { join } = new Path()
    expect(join('a/b', 'c/..')).toBe('a/b')
  })
})

describe('resolve', () => {
  it('relative path only', () => {
    const { resolve } = new Path()
    expect(resolve('a', 'b/c', 'd/e/f')).toBe('/working/directory/a/b/c/d/e/f')
  })

  it('with one absolute path', () => {
    const { resolve } = new Path()
    expect(resolve('a/b', 'c', '/d/e/f')).toBe('/d/e/f')
  })
})
