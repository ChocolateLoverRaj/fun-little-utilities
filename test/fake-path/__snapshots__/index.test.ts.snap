// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isAbsolute 1`] = `
Object {
  "absolute": Array [
    "/",
    "C:\\\\",
    "/a/b/c",
    "C:\\\\a\\\\b",
  ],
  "relative": Array [
    ".",
    "..",
    "a/b/c",
    "./a/b",
    "../a/b",
  ],
}
`;

exports[`normalize edge cases: empty 1`] = `
"
        '' => '.'
     './.' => '.'
    './/.' => '.'
   '././.' => '.'
      './' => '.'
     './/' => '.'
"
`;

exports[`normalize edge cases: root 1`] = `
"
       '/' => '/'
      '//' => '/'
     '///' => '/'
      '/.' => '/'
    '/./.' => '/'
   '/.//.' => '/'
  '/././.' => '/'
    '//./' => '/'
    '/.//' => '/'
"
`;
